[clang-tidy] Add a release note about unchecked-optional-access smart pointer caching

jvoung opened the pull request:
With caching added in https://github.com/llvm/llvm-project/pull/120249, the `IgnoreSmartPointerDereference` option shouldn't be needed anymore.

Other caching also added earlier: https://github.com/llvm/llvm-project/pull/112605

jvoung committed:
[clang-tidy] Add a release note about unchecked-optional-access smart pointer caching

With caching added in https://github.com/llvm/llvm-project/pull/120249,
inform in notes that the `IgnoreSmartPointerDereference` option shouldn't
be needed anymore.

Other caching also added earlier:
https://github.com/llvm/llvm-project/pull/112605

jvoung committed:
Clean up text

HerrCai0907 reviewed (approved):
does check doc need to be changed also?

jvoung committed:
Also update checker docs a bit

We updated some earlier, but not the later section about recommending
binding to a local variable. Also did not update w.r.t. smart pointer
like APIs.

jvoung commented:
> does check doc need to be changed also?

Good question -- I don't see the option IgnoreSmartPointerDereference advertised in the check docs.

Otherwise, we partially updated the check docs about earlier accessor caching. Updated more to mention smart pointer like APIs. Also previously updated internal code comments about IgnoreSmartPointerDereference in the `UncheckedOptionalAccessModelOptions` struct.

EugeneZelenko reviewed (commented):
Could you please also move `readability-identifier-naming` entry before `readability-implicit-bool-conversion` and place `:program:run-clang-tidy.py` entry after `:program:clang-tidy.py`?

jvoung commented:
> program:run-clang-tidy

Sure thing --

Done separately in https://github.com/llvm/llvm-project/pull/122475 (I kept the "Improved" cluster together, vs moving `run-clang-tidy` to be after the "Removed" block including the table)

For `readability-identifier-naming` it looks like @HerrCai0907 already reordered that in https://github.com/llvm/llvm-project/commit/c01ffab6e1dac13a344a06677f413e55073d0580

jvoung committed:
Simplify note a bit more

jvoung commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unchecked-optional-access.rst:
```diff
@@ -81,10 +81,12 @@ Exception: accessor methods
+methods and smart pointer-like APIs (non ``const`` overloads of ``*`` when
+there is a parallel ``const`` overload) are treated as accessors. Note that
+this is not guaranteed to be safe -- but, it is widely used (safely) in
+practice, and so we have chosen to treat it as generally safe. Calls to non
```
Done

jvoung committed:
Merge branch 'main' into release_notes_about_smartptr_options

5chmidti commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -239,6 +239,10 @@ Changes in existing checks
   ``bsl::optional`` and ``bdlb::NullableValue`` from
   <https://github.com/bloomberg/bde>_.
+  Fixed false positives from smart pointer accessors repeated in checking
+  ``has_value`` and accessing ``value``, by caching the locations returned
+  by the accessors. The option `IgnoreSmartPointerDereference` should no
```
The part `by caching the locations returned by the accessors` is IMO not really something a user needs wants to know, as it is an implementation detail.

jvoung committed:
Remove part about caching, and clarify that option will be removed.

jvoung committed:
Merge branch 'main' into release_notes_about_smartptr_options

jvoung commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -232,6 +232,10 @@ Changes in existing checks
+  Fixed false positives from smart pointer accessors repeated in checking
+  ``has_value`` and accessing ``value``, by caching the locations returned
+  by the accessors. The option `IgnoreSmartPointerDereference` should no
+  longer be needed.
```
I think it can be removed. By "maybe could be removed", did you mean to actually remove the option (together with this change or as a separate change), or to clarify in the release note that it could be removed in the future/soon?

I wasn't sure if it was okay to remove a flag before advertising that it isn't needed.  (though It looks like right now clang-tidy just ignores options that aren't available, instead of failing with an error)?

jvoung committed:
Merge branch 'main' into release_notes_about_smartptr_options

jvoung committed:
Fix merge
